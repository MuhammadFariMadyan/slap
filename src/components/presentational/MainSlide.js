import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import Textbox from "./Textbox";

const Container = styled.svg`
  width: 640px;
  height: 480px;
  border: 1px solid #ccc;
  background: white;
  margin: auto;
  box-shadow: 0px 0px 2px 1px rgba(0, 0, 0, 0.2);
`;

const MainSlide = ({
  contents,
  active,
  cursor,
  onClick,
  onChangeContent,
  onClickContent,
  onDoubleClickContent,
  onMouseMove,
  isContentEditable
}) => (
  <Container onClick={onClick} onMouseMove={onMouseMove} cursor={cursor}>
    {contents.map(content => (
      <Textbox
        key={content.id}
        width={320}
        height={64}
        {...content.position}
        value={content.content}
        onChange={event => onChangeContent(event, content)}
        onClick={event => onClickContent(event, content)}
        onDoubleClick={event => onDoubleClickContent(event, content)}
        isActive={active === content.id}
        isEditable={isContentEditable}
      />
    ))}
  </Container>
);

MainSlide.propTypes = {
  contents: PropTypes.array,
  active: PropTypes.string,
  cursor: PropTypes.string,
  onClick: PropTypes.func,
  onChangeContent: PropTypes.func,
  onClickContent: PropTypes.func,
  onDoubleClickContent: PropTypes.func,
  onMouseMove: PropTypes.func,
  isContentEditable: PropTypes.bool
};

export default MainSlide;
