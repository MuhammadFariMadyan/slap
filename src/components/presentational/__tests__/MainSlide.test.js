import React from "react";
import { shallow } from "enzyme";

import MainSlide from "../MainSlide";

jest.mock("../Textbox", () => "Textbox");

describe("MainSlide", () => {
  const setup = propOverrides => {
    const props = {
      contents: [
        {
          id: "content-1",
          slideID: "slide-1",
          projectID: "project-1",
          type: "text",
          content: "Hello World",
          position: {
            x: 32,
            y: 32
          }
        },
        {
          id: "content-2",
          slideID: "slide-1",
          projectID: "project-1",
          type: "text",
          content: "Stay hungry",
          position: {
            x: 128,
            y: 96
          }
        }
      ],
      onClick: jest.fn(),
      onChangeContent: jest.fn(),
      onClickContent: jest.fn(),
      onDoubleClickContent: jest.fn(),
      onMouseMove: jest.fn(),
      ...propOverrides
    };

    const wrapper = shallow(<MainSlide {...props} />);

    return {
      wrapper,
      props
    };
  };

  it("should renders default", () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it("should renders with active content", () => {
    const { wrapper } = setup({ active: "content-2" });
    const textbox = wrapper.find("Textbox");

    expect(textbox.at(1).props().isActive).toEqual(true);
  });

  it("should renders with editable content", () => {
    const { wrapper } = setup({
      active: "content-2",
      isContentEditable: true
    });
    const textbox = wrapper.find("Textbox");

    expect(textbox.at(1).props().isActive).toEqual(true);
    expect(textbox.at(1).props().isEditable).toEqual(true);
  });

  it("should calls 'onClick' callback while clicked", () => {
    const { wrapper, props } = setup();

    wrapper.simulate("click");

    expect(props.onClick).toBeCalled();
  });

  it("should calls 'onChangeContent' callback while content changed", () => {
    const { wrapper, props } = setup();
    const event = {
      target: {
        value: "Hello World"
      }
    };

    props.contents.forEach((content, index) => {
      wrapper
        .find("Textbox")
        .at(index)
        .simulate("change", event);

      expect(props.onChangeContent).toBeCalledWith(event, content);
    });
  });

  it("should calls 'onClickContent' callback while content clicked", () => {
    const { wrapper, props } = setup();
    const event = { stopPropagation: () => {} };

    props.contents.forEach((content, index) => {
      wrapper
        .find("Textbox")
        .at(index)
        .simulate("click", event);

      expect(props.onClickContent).toBeCalledWith(event, content);
    });
  });

  it("should calls 'onDoubleClickContent' callback while content double clicked", () => {
    const { wrapper, props } = setup();
    const event = { stopPropagation: () => {} };

    props.contents.forEach((content, index) => {
      wrapper
        .find("Textbox")
        .at(index)
        .simulate("dblclick", event);

      expect(props.onDoubleClickContent).toBeCalledWith(event, content);
    });
  });

  it("should calls 'onMouseMove' callback while mouse moved", () => {
    const { wrapper, props } = setup();

    wrapper.simulate("mousemove");

    expect(props.onMouseMove).toBeCalled();
  });
});
