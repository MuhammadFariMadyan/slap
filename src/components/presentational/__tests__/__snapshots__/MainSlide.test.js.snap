// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainSlide should renders default 1`] = `
<styled.svg
  cursor="pointer"
  innerRef={null}
  onClick={[Function]}
  onMouseMove={[Function]}
>
  <Textbox
    height={64}
    isActive={false}
    key="content-1"
    onChange={[Function]}
    onClick={[Function]}
    onDoubleClick={[Function]}
    value="Hello World"
    width={320}
    x={32}
    y={32}
  />
  <Textbox
    height={64}
    isActive={false}
    key="content-2"
    onChange={[Function]}
    onClick={[Function]}
    onDoubleClick={[Function]}
    value="Stay hungry"
    width={320}
    x={128}
    y={96}
  />
</styled.svg>
`;
